### PR TITLE
Last Python 3.11.x release.

### LINKED ISSUE #55: Drop all binary wheels.
From https://github.com/chevah/server/pull/6613:
```
The imporant part is to build the next version of Pythia without binary wheels.
We can have those wheels installed in chevah/server.
```

One this is implemented, we should use it in chevah/server and do some cleanup as right now in chevah/server we have a few workarounds https://github.com/chevah/server/issues/6639

### PR DESCRIPTION
**Pull request by dumol**
Scope
-----

While working on #57, it became obvious it would be little work to also release a last set of Python 3.11 packages. 

Changes
-------

Updates Python to 3.11.9, the last bugfix release for 3.11.

OpenSSL on non-Windows platforms (for Python's `ssl` module) now uses the 3.0.x branch, in line with Windows.

Updated all the other libs:
  * libffi to 3.4.6
  * zlib to 1.3.1.
  * xz to 5.6.2
  * SQLite to 3.46.0.

Platforms other than Windows use `ensurepip` to install `pip`.
 
Updated `pip` to 24.0.

Removed all binary-only wheels. Fixes #55.

Updated remaining Python modules:
  * `setuptools` to 70.0.0
  * `psutil` to 5.9.8
  * `pycparser` to 2.22.

**Drive-by changes**:
  * Added an option for `safety` IDs to ignore
  * Updated GitHub workflows to run `tmate` when retrying with debug enabled
  * Updated GitHub workflows to use the new GitHub m1-based macOS runners
  * Build for macOS 11 on Apple Silicon (still for 10.13 on Intel)
  * Prevent unwanted GCC optimizations
  * Use `CC` and `CXX` from env if defined.
  * Accept musl git versions.

Testing
-------

Automated.

### CODE REVIEW

File: .github/workflows/bare.yaml [hunk 1]
```diff
@@ -14,19 +14,11 @@ concurrency:
...
-  TMATE_DEBUG: 'no'
-  NGROK_DEBUG: 'no'
-
-
 jobs:
   windows:
     # The type of runner that the job will run on
     runs-on: windows-latest
-    timeout-minutes: 15
+    timeout-minutes: ${{ matrix.tests.debug == 'yes' && 45 || 15 }}
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:21]: do we have a matrix here ?

File: .github/workflows/bare.yaml [hunk 2]
```diff
@@ -60,9 +52,8 @@ jobs:
     - name: Upload testing package
       run: bash -c './publish_dist.sh ; rm priv_key'
 
-    # Command line debugging through Tmate. v3 works on Windows too.
     - name: Tmate debug on failure
-      if: failure() && env.TMATE_DEBUG == 'yes'
+      if: ${{ failure() && !cancelled() && runner.debug }}
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:56]: One thing that I did, is trigger tmage only for manual `debug` re-runs.

You can trigger a re-run with debug enabled only for a single job

And another idea is to move the `tmate` step before the test run
ON a manual re-run you can manually re-run the failed test, and don't have to wait that long before tmate is available
  
```
  - name: Test Pythia
      run: bash ./build.sh test
```

```suggestion
      if: ${{ !cancelled() && runner.debug }}
```

File: .github/workflows/bare.yaml [hunk 3]
```diff
@@ -86,8 +77,9 @@ jobs:
 
 
   macos:
-    runs-on: macos-latest
-    timeout-minutes: 30
+    # For macOS on Intel at GitHub.
+    runs-on: macos-13
+    timeout-minutes: ${{ matrix.tests.debug == 'yes' && 90 || 30 }}
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:82]: do we have a matrix here ?

File: .github/workflows/bare.yaml [hunk 4]
```diff
@@ -131,7 +123,8 @@ jobs:
 
 
   macos-m1:
-    runs-on: m1
+    # For macOS on Apple Silicon at GitHub.
+    runs-on: macos-latest
     timeout-minutes: ${{ matrix.tests.debug == 'yes' && 90 || 30 }}
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:128]: do we have a matrix here ?

File: build.conf [hunk 5]
```diff
@@ -6,50 +6,47 @@ BUILD_DIR="build"
...
-PYWIN32_VERSION="306"
+PSUTIL_VERSION="5.9.8"
 
 # pycparser is explicitly installed to work around setuptools auto dependencies.
-PYCPARSER_VERSION="2.21"
+PYCPARSER_VERSION="2.22"
 
 # safety version is pinned to avoid picking older ones from our PyPI server.
-SAFETY_VERSION="2.3.5"
+SAFETY_VERSION="3.2.0"
```
Comment (adiroiban) [inline build.conf:49]: pip should not pick older version.

but we want to pin safety, to make sure the deps of the test environment are controlled

File: build.conf [hunk 6]
```diff
@@ -64,15 +61,14 @@ BUILD_SQLITE="yes"
 PIP_INDEX_URL="https://bin.chevah.com:20443/pypi/simple"
 # Array of default Python modules installed using pip.
 PIP_LIBRARIES=(\
-    cffi=="$CFFI_VERSION" \
-    bcrypt=="$BCRYPT_VERSION" \
     psutil=="$PSUTIL_VERSION" \
```
Comment (adiroiban) [inline build.conf:64]: I think that we can also remove psutil for pythia.

File: build.conf [hunk 7]
```diff
@@ -6,50 +6,47 @@ BUILD_DIR="build"
...
+OPENSSL_VERSION="3.0.14"
 # Use the version of the "sqlite-autoconf-VERSION.tar.gz" upstream download.
 # To get its SHA3-256 signature: "openssl dgst -sha3-256 sqlite-autoconf-*".
 # When updating this, also update the year in src/sqlite/chevahbs, if needed.
-SQLITE_VERSION="3440200"
+SQLITE_VERSION="3460000"
 
 # If latest get-pip.py stops working on older platforms, use something like:
 # https://web.archive.org/web/20220122185015/https://bootstrap.pypa.io/.
 BOOTSTRAP_GET_PIP="https://bootstrap.pypa.io/get-pip.py"
```
Comment (adiroiban) [inline build.conf:35]: With latest Python,  I don't think that this is needed.

You should be able to just do `python -m ensurepip --upgrade`

File: build.conf [hunk 8]
```diff
@@ -6,50 +6,47 @@ BUILD_DIR="build"
...
 
 # Python modules versions to be used everywhere possible.
 # Latest cryptography/bcrypt require Rust, use only wheels for them.
-CFFI_VERSION="1.16.0"
-BCRYPT_VERSION="4.1.1"
-PSUTIL_VERSION="5.9.6"
-SETPROCTITLE_VERSION="1.3.3"
-MARKUPSAFE_VERSION="2.1.3"
-CHARSETNORMALIZER_VERSION="3.3.2"
-PYWIN32_VERSION="306"
```
Comment (adiroiban) [inline build.conf:46]: it would help to continue to distribute pywin32 with pythia

File: build.conf [hunk 9]
```diff
@@ -6,50 +6,47 @@ BUILD_DIR="build"
...
+PSUTIL_VERSION="5.9.8"
```
Comment (adiroiban) [inline build.conf:43]: why not also remove psutil ?

**Review by adiroiban — APPROVED** [review]
thanks. looks good. only minor comment.

I think we can merge this and then I will create a PR in chevah/server to give it a try.

Thanks again

File: build.conf (hunk 6 above)
Comment (dumol) [reply build.conf:64]: Unfortunately, there is no musl wheel for `psutil`.

File: build.conf (hunk 8 above)
Comment (dumol) [reply build.conf:46]: Fixed with 47ff497.

File: build.conf (hunk 9 above)
Comment (dumol) [reply build.conf]: There's no `psutil` wheel for `linux_musl`, so we have to build it.

File: build.conf (hunk 7 above)
Comment (dumol) [reply build.conf]: Fixed through a3c646f.

File: build.conf (hunk 5 above)
Comment (dumol) [reply build.conf:44]: Fixed comment.

File: .github/workflows/bare.yaml (hunk 1 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: What do you suggest? I've tried using `runner.debug` but it wouldn't work: _Unrecognized named-value: 'runner'_.

I've also tried simply `test.debug`, to no avail.

File: .github/workflows/bare.yaml (hunk 2 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: Committed suggestion through 22ea68f5.

File: build.conf (hunk 7 above)
Comment (dumol) [reply build.conf]: Unfortunately, this is not a solution on Windows, where the upstream package we use doesn't seem to include `ensurepip`:
```
## Bootstrapping pip... ##
D:\a\pythia\pythia\build\python3.11-windows-x64\lib\python.exe: No module named ensurepip
```

Comment (adiroiban) [reply build.conf]: Thanks. Not sure why ensurepip is not available on Windows :(

**Comment by adiroiban** [comment]
> Unfortunately, there is no musl wheel for psutil.

Maybe just add a comment to inform why we have `psutil`.

In the future updates, we can review the pypi files for psutil.. maybe musl will be added in the future and we can remove this.

I see that nobody asked for `musl`  package https://github.com/giampaolo/psutil/issues?q=is%3Aissue+is%3Aopen+musl

Maybe we can create a PR and add support for musl

It should be possible to build it via Docker

File: .github/workflows/bare.yaml (hunk 1 above)
Comment (adiroiban) [reply .github/workflows/bare.yaml]: The `runner` is not available at the "jobs->JOB_ID" level.
It is only avalable at the `jobs->JOB_ID->steps->STEP_ID` level

Just leave a big timeout here, and make sure each step has a lower timeout, with the exception of the debug step

**Comment by adiroiban** [comment]
We will see how fast will be the migration to 3.13 .. so this might not be the last release.

But the focus should be on 3.13

the Python ecosystem is slowly moving toward 3.13 ... recently we got [cffi on 3.13](https://github.com/python-cffi/cffi/releases/tag/v1.17.0rc1)

**Comment by adiroiban** [comment]
We can release it this version and I will include it in the next chevah/server update

File: build.conf (hunk 7 above)
Comment (dumol) [reply build.conf:36]: The embeddable package on Windows is cut to the bone.

Went back to using `get-pip.py`, but only on Windows: https://github.com/chevah/pythia/pull/59/files/cbd3bc16b1cbb257d59f0e2cf5dedff1ca7ef039..4d6db81c9bd9df5161200c5ae8b70ac6f5bdd107

File: .github/workflows/bare.yaml (hunk 1 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: Fixed in f9d6da61930cc40e6f555359cb3e02ead8fe7130.

File: .github/workflows/bare.yaml (hunk 4 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: Fixed in f9d6da6.

File: .github/workflows/bare.yaml (hunk 3 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: Fixed in f9d6da6.

**Comment by dumol** [comment]
> We will see how fast will be the migration to 3.13 .. so this might not be the last release.

It's definitely the last Python 3.11 release. At least, that's what upstream says at https://www.python.org/downloads/release/python-3119/:  _This is the last bugfix release of the Python 3.11 series._

**Comment by dumol** [comment]
@adiroiban, FYI, I've seen GCC optimizing the build on Linux for whatever processor it detected on GitHub's VM, so I've added the `-mtune=generic` flag on Linux.

Also, on macOS, we now build for macOS 11.0 on Apple Silicon, which is the oldest macOS version with support for ARM64. (We still build for 10.3 on Intel hardware, as long as this works without much work.)

On Windows, we don't build anything at all.

**Pull request merged**